Add trim support to the sg ioengine

vincentkfu opened the pull request:
This is a follow-up to #590 

Jens, please consider this pull request.

These patches add trim support to the sg ioengine via the SCSI unmap command. Following the established sg pattern, unmap commands for block devices are carried out synchronously. For character devices, however, it is possible to submit multiple ranges with a single unmap command by setting iodepth_batch.

@sitsofe, I considered adding an option to use WRITE SAME but using unmap was my first priority since that was what was requested by the group that this feature was developed for. Note that it would be pretty easy to add an option so that write workloads with sg_write_mode=same carry out trims.

axboe commented on engines/sg.c:
```diff
@@ -3,6 +3,39 @@
+ *
+ *				trim operations are queued in queue() and submitted in commit()
+ *			T	called in commit()		submit called in commit()	set in commit()
+ *
```
I like the detailed explanations, but please keep it within 80 chars. It's unreadable in my term.

axboe commented on engines/sg.c:
```diff
@@ -21,6 +54,8 @@ enum {
 	FIO_SG_WRITE_SAME	= 3
 };
 
+static struct ioengine_ops ioengine;
```
Nope... See below.

axboe commented on engines/sg.c:
```diff
@@ -383,65 +501,120 @@ static int fio_sgio_prep(struct thread_data *td, struct io_u *io_u)
 	else if (hdr->status) {
 		io_u->resid = hdr->resid;
 		io_u->error = EIO;
+	} else if (ioengine.commit != NULL) {
```
This is really backwards. td->io_ops->commit if you must check.

axboe commented on ioengines.c:
```diff
@@ -350,7 +350,9 @@ enum fio_q_status td_io_queue(struct thread_data *td, struct io_u *io_u)
-	if (!td->io_ops->commit || io_u->ddir == DDIR_TRIM) {
+	if (!td->io_ops->commit ||
+	    (io_u->ddir == DDIR_TRIM &&
+             !td_ioengine_flagged(td, FIO_ASYNCTRIM))) {
```
Instead of adding FIO_ASYNCTRIM, we should stop special casing trim here.

vincentkfu commented:
@axboe, I have updated the branch based on your comments. Here are the main changes:

- reformatted the io accounting comments
- now use td->io_ops->commit to check if commit is null
- eliminated FIO_ASYNCTRIM
- added patch with HOWTO and manpage changes
- changed sgunmap-test.py to be more informative when an assertion fails
- add more detail to the sg error string

axboe commented:
Pulled, thanks Vincent!
